examples/mem-correctness-test/src/main.rs

```diff
@@ -0,0 +1,267 @@
+
+const NUM_OPERATION_CHOICES: usize = 4;
+const MAGIC_NUM_TO_CHECK: u32 = 0xdeadbeef;
+static heap_allocated: AtomicUsize = AtomicUsize::new(0);
+
+/* Set of configurable parameters. These will adjusted as necessary while
+ * recording the performance numbers
+ */
+const NUM_CPUS: usize = 64;
+const PER_THREAD_PER_BUFFER_MAX_SIZE: usize = (4 * TO_MB);
+const MAX_BUFF_NUM_PER_THREAD: usize = (32);
+const MAX_BUFFER_CHECKS_PER_THREAD_ITERATION: usize = 8;
+const MAX_INDEX_CHECKS_PER_BUFFER: usize = 1 << 16;
+const MAX_ALLOWED_FREE_HEAP_PERCENTAGE: f64 = 80.0;
+const MAX_ITERATIONS_PER_THREAD: usize = 100000;
+
+#[inline(always)]
+pub fn image_base() -> u64 {
+    let base: u64;
+    unsafe {
+        asm!(
+            "lea IMAGE_BASE(%rip), {}",
+            lateout(reg) base,
+            options(att_syntax, nostack, preserves_flags, nomem, pure),
+        )
+    };
+    base
+}
+
+#[inline(always)]
+pub(crate) unsafe fn rel_ptr_mut<T>(offset: u64) -> *mut T {
+    (image_base() + offset) as *mut T
+}
+
+/* Returns the base memory address of the heap */
+pub(crate) fn heap_base() -> *const u8 {
+    unsafe { rel_ptr_mut(HEAP_BASE) }
+}
+
+/* Returns the size of the heap */
+pub(crate) fn heap_size() -> usize {
+    unsafe { HEAP_SIZE }
+}
+
+fn update_occupied_heap_size_on_delete(size: usize) {
+    heap_allocated.fetch_sub(size, Ordering::SeqCst);
+}
+
+fn update_occupied_heap_size_on_addition(size: usize) {
+    heap_allocated.fetch_add(size, Ordering::SeqCst);
+}
+
+fn get_occupied_heap_size() -> usize {
+    heap_allocated.load(Ordering::SeqCst)
+}
+
+fn get_free_heap_size_in_bytes() -> usize {
+    heap_size() - get_occupied_heap_size()
+}
+
+fn get_free_heap_percentage() -> f64 {
+    (get_free_heap_size_in_bytes() as f64 / heap_size() as f64) * 100.0
+}
+
+fn get_random_num(start: usize, end: usize) -> usize {
+    let mut rng = rand::thread_rng();
+    rng.gen_range(start..=end)
+}
+
+fn wait_per_thread(pair_clone: Arc<(Mutex<bool>, Condvar)>) {
+    let (lock, cvar) = &*pair_clone;
+    let mut started = lock.lock().unwrap();
+    while !*started {
+        started = cvar.wait(started).unwrap();
+    }
+    drop(started);
+}
+
+fn wakeup_all_child_threads(pair_clone: Arc<(Mutex<bool>, Condvar)>) {
+    let (lock, cvar) = &*pair_clone;
+    let mut started = lock.lock().unwrap();
+    *started = true;
+    cvar.notify_all();
+    drop(started);
+}
+
+fn traverse_and_check_buffer(buf: &Vec<u32>) {
+    let mut ptr = 0;
+    let num_indices_checks = get_random_num(1, MAX_INDEX_CHECKS_PER_BUFFER);
+    for i in 1..=num_indices_checks {
+        /* Check for random indices and number of such indices is  num_indices_checks
+         * Please note that depending on the number random number generator, we
+         * can check for the same index multiple times. We could have checked
+         * for all the indices but that would be too time consuming
+         */
+        if (buf[get_random_num(0, buf.len() - 1)] != MAGIC_NUM_TO_CHECK) {
```

Comment:
Maybe as a future improvement, it may be a good idea to fill the buffer with pseudo-random data that can be easily checked, rather than by repeating the same bytes over and over. I've worked on memory allocators in the past that had bugs that manifested as data being copied to the wrong location in memory. These bugs were not caught by a test similar to this one that repeated the same magic number repeatedly, because all of the magic numbers were the same. When we changed the test to use pseudo-random (but calculatable) sequences, the test was able to catch the bug.